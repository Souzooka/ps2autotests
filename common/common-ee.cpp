#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
