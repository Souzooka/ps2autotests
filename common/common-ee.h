#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
