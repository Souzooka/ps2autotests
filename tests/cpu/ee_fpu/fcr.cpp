#include <common-ee.h>
#include <string.h>

void __attribute__((noinline)) printfcrs(const char *title, int all) {
	unsigned int fcrs[32];
	memset(fcrs, 0xCC, sizeof(fcrs));

#define ASM_READ_FCR(R) \
		"lui $t0, 0\n" \
		"cfc1 $t0, $" #R "\n" \
		"nop\n" \
		"sw $t0, 4 * " #R "(%0)\n"

	asm volatile (
		".set noreorder\n"

		ASM_READ_FCR(0)
		ASM_READ_FCR(1)
		ASM_READ_FCR(2)
		ASM_READ_FCR(3)
		ASM_READ_FCR(4)
		ASM_READ_FCR(5)
		ASM_READ_FCR(6)
		ASM_READ_FCR(7)
		ASM_READ_FCR(8)
		ASM_READ_FCR(9)
		ASM_READ_FCR(10)
		ASM_READ_FCR(11)
		ASM_READ_FCR(12)
		ASM_READ_FCR(13)
		ASM_READ_FCR(14)
		ASM_READ_FCR(15)
		ASM_READ_FCR(16)
		ASM_READ_FCR(17)
		ASM_READ_FCR(18)
		ASM_READ_FCR(19)
		ASM_READ_FCR(20)
		ASM_READ_FCR(21)
		ASM_READ_FCR(22)
		ASM_READ_FCR(23)
		ASM_READ_FCR(24)
		ASM_READ_FCR(25)
		ASM_READ_FCR(26)
		ASM_READ_FCR(27)
		ASM_READ_FCR(28)
		ASM_READ_FCR(29)
		ASM_READ_FCR(30)
		ASM_READ_FCR(31)

		".set    reorder\n"
		: : "r"(fcrs) : "t0"
	);

	printf("%s:\n  ", title);
	if (all) {
		int i;
		for (i = 0; i < 31; i++) {
			printf("fcr%d: %08x, ", i, fcrs[i]);
		}
	} else {
		printf("fcr0: %08x, ", fcrs[0]);
	}
	printf("fcr31: %08x\n", fcrs[31]);
}

void __attribute__((noinline)) setfcr0(u32 v) {
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"ctc1    %0, $0\n"
		"nop\n"

		".set    reorder\n"
		".set    macro\n"
		: : "r"(v)
	);
}

void __attribute__((noinline)) updatefcr31(u32 v) {
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"or      $v0, $0, %0\n"
		"ctc1    $v0, $31\n"
		"nop\n"

		".set    reorder\n"
		".set    macro\n"
		: : "r"(v) : "v0"
	);
}

static float sqrtneg() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"lui     $v0, 0xBF80\n" // 0xBF800000 = -1
		"mtc1    $v0, %0\n"
		"nop\n"
		"sqrt.s  %0, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res) : : "v0"
	);
	return res;
}

static float zerodivzero() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"mtc1    $0, %0\n"
		"nop\n"
		"div.s   %0, %0, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res)
	);
	return res;
}

static float numdivzero() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"lui     $v0, 0x3F80\n" // 0x3F800000 = 1
		"mtc1    $v0, $f1\n"
		"mtc1    $0, %0\n"
		"nop\n"
		"div.s   %0, $f1, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res) : : "$f1"
	);
	return res;
}

static float nanmath() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"lui     $v0, 0x7F80\n" // 0x7F800001 = NAN
		"ori     $v0, 0x0001\n"
		"mtc1    $v0, %0\n"
		"nop\n"
		"add.s   %0, %0, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res) : : "v0"
	);
	return res;
}

static float fltoverflow() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"lui     $v0, 0x7F7F\n" // 0x7F7FFFFF = FLT_MAX
		"ori     $v0, 0xFFFF\n"
		"mtc1    $v0, %0\n"
		"nop\n"
		"mul.s   %0, %0, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res) : : "v0"
	);
	return res;
}

static float fltunderflow() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"lui     $v0, 0x0080\n" // 0x00800000 = FLT_MIN
		"mtc1    $v0, $f1\n"
		"lui     $v0, 0x4040\n" // 0x40400000 = 3
		"mtc1    $v0, %0\n"
		"nop\n"
		"div.s   %0, $f1, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res) : : "v0", "$f1"
	);
	return res;
}

static float fltinexact() {
	float res;
	asm volatile (
		".set    noreorder\n"
		".set    nomacro\n"

		"lui     $v0, 0x3F80\n" // 0x3F800000 = 1
		"mtc1    $v0, $f1\n"
		"lui     $v0, 0x4040\n" // 0x40400000 = 3
		"ori     $v0, 0xFFFF\n"
		"mtc1    $v0, %0\n"
		"nop\n"
		"div.s   %0, $f1, %0\n"

		".set    reorder\n"
		".set    macro\n"
		: "=f"(res) : : "v0", "$f1"
	);
	return res;
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	printfcrs("Initial", 1);

	setfcr0(0xDEADBEEF);
	printfcrs("Update fcr0", 0);
	updatefcr31(0x00000003);
	printfcrs("Update rounding mode (RM)", 0);
	updatefcr31(0x0000007C);
	printfcrs("Update flags", 0);
	printfcrs("All regs", 1);
	updatefcr31(0x00000F80);
	printfcrs("Update enables", 0);
	updatefcr31(0x0001F000);
	printfcrs("Update cause", 0);
	updatefcr31(0x00020000);
	printfcrs("Update unimplemented (E)", 0);
	updatefcr31(0x01000000);
	printfcrs("Update flushing (FS)", 0);
	updatefcr31(0x00400000);
	printfcrs("Update flushing (FO)", 0);
	updatefcr31(0x00200000);
	printfcrs("Update flushing (FN)", 0);
	updatefcr31(0x00800000);
	printfcrs("Update FCC", 0);
	updatefcr31(0xFE000000);
	printfcrs("Update FCC1-7", 0);
	updatefcr31(0x001C0000);
	printfcrs("Update unknown", 0);

	printf("\nFlag situations:\n");

	float f1;

	updatefcr31(0x00000000);
	f1 = sqrtneg();
	printfcrs("sqrt(-1)", 0);
	printf("  result = %f\n", f1);

	updatefcr31(0x00000000);
	f1 = zerodivzero();
	printfcrs("Divide zero by zero", 0);
	printf("  result = %f\n", f1);

	updatefcr31(0x00000000);
	f1 = numdivzero();
	printfcrs("Divide one by zero", 0);
	printf("  result = %f\n", f1);

	updatefcr31(0x00000000);
	f1 = nanmath();
	printfcrs("NAN math", 0);
	printf("  result = %f\n", f1);

	updatefcr31(0x00000000);
	f1 = fltoverflow();
	printfcrs("Overflow", 0);
	printf("  result = %f\n", f1);

	updatefcr31(0x00000000);
	// Can't seem to make underflow happen.
	f1 = fltunderflow();
	printfcrs("Underflow", 0);
	printf("  result = %f\n", f1);

	updatefcr31(0x00000000);
	f1 = fltinexact();
	printfcrs("Inexact", 0);
	printf("  result = %f\n", f1);

	printf("-- TEST END\n");

	return 0;
}
