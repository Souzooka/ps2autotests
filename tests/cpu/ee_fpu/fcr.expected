-- TEST BEGIN
Initial:
  fcr0: 00002e30, fcr1: 00002e30, fcr2: 00002e30, fcr3: 00002e30, fcr4: 00002e30, fcr5: 00002e30, fcr6: 00002e30, fcr7: 00002e30, fcr8: 00002e30, fcr9: 00002e30, fcr10: 00002e30, fcr11: 00002e30, fcr12: 00002e30, fcr13: 00002e30, fcr14: 00002e30, fcr15: 00002e30, fcr16: 01000001, fcr17: 01000001, fcr18: 01000001, fcr19: 01000001, fcr20: 01000001, fcr21: 01000001, fcr22: 01000001, fcr23: 01000001, fcr24: 01000001, fcr25: 01000001, fcr26: 01000001, fcr27: 01000001, fcr28: 01000001, fcr29: 01000001, fcr30: 01000001, fcr31: 01000001
Update fcr0:
  fcr0: 00002e30, fcr31: 01000001
Update rounding mode (RM):
  fcr0: 00002e30, fcr31: 01000001
Update flags:
  fcr0: 00002e30, fcr31: 01000079
All regs:
  fcr0: 00002e30, fcr1: 00002e30, fcr2: 00002e30, fcr3: 00002e30, fcr4: 00002e30, fcr5: 00002e30, fcr6: 00002e30, fcr7: 00002e30, fcr8: 00002e30, fcr9: 00002e30, fcr10: 00002e30, fcr11: 00002e30, fcr12: 00002e30, fcr13: 00002e30, fcr14: 00002e30, fcr15: 00002e30, fcr16: 01000079, fcr17: 01000079, fcr18: 01000079, fcr19: 01000079, fcr20: 01000079, fcr21: 01000079, fcr22: 01000079, fcr23: 01000079, fcr24: 01000079, fcr25: 01000079, fcr26: 01000079, fcr27: 01000079, fcr28: 01000079, fcr29: 01000079, fcr30: 01000079, fcr31: 01000079
Update enables:
  fcr0: 00002e30, fcr31: 01000001
Update cause:
  fcr0: 00002e30, fcr31: 0101c001
Update unimplemented (E):
  fcr0: 00002e30, fcr31: 01020001
Update flushing (FS):
  fcr0: 00002e30, fcr31: 01000001
Update flushing (FO):
  fcr0: 00002e30, fcr31: 01000001
Update flushing (FN):
  fcr0: 00002e30, fcr31: 01000001
Update FCC:
  fcr0: 00002e30, fcr31: 01800001
Update FCC1-7:
  fcr0: 00002e30, fcr31: 01000001
Update unknown:
  fcr0: 00002e30, fcr31: 01000001

Flag situations:
sqrt(-1):
  fcr0: 00002e30, fcr31: 01020041
  result = 1.000000
Divide zero by zero:
  fcr0: 00002e30, fcr31: 01020041
  result = NaN
Divide one by zero:
  fcr0: 00002e30, fcr31: 01010021
  result = NaN
NAN math:
  fcr0: 00002e30, fcr31: 01008011
  result = NaN
Overflow:
  fcr0: 00002e30, fcr31: 01008011
  result = NaN
Underflow:
  fcr0: 00002e30, fcr31: 01000001
  result = 0.000000
Inexact:
  fcr0: 00002e30, fcr31: 01000001
  result = 0.331606
-- TEST END
